templated-this.md

```diff
@@ -1,151 +1,315 @@
+from inside its function call operator, the implementer has a dilemma: incur a
+copy, or use a move, and impose a requirement that is unenforceable in code and
+sets hidden traps for the unwary.
+
+
+Design Decisions
+----------------
+   
+In addition to solving the existing problems, desirable properties of a solution
+are:
+
+* It should work the same way for member functions and lambda expressions.
+
+* It should work like existing practice as much as possible, while adding as
+  little extra syntax as possible. Using `auto&& this` follows existing patterns
+  for dealing with forwarding references. Optionally adding "this" as the first
+  parameter fits with many programmers' mental model of the `this` pointer being
+  the first parameter to member functions "under the hood" and is comparable to
+  usage in other languages, e.g. Python.
+
+* It should avoid adding extraneous syntax to function declarations. In a world
+  where every function is fast becoming tagged with attributes, constexpr,
+  noexcept specifications, and (this proposal notwithstanding) reference
+  qualifiers, readable, noise-free function declarations are to be prized.
+
+TBD: more here.
+
+
+Proposed Solution
+-----------------
+
+We propose the ability to add an optional first parameter to any member function
+of a class `T`, taking the form `T [const] [volatile] [&|&&] this`.
+
+To facilitate use in generic lambda expressions, this may also be formulated as
+`auto [const] [volatile] [&|&&] this`.
+
+In all cases, the value-category of `this` inside the member function is exactly
+what the existing parameter rules already imply.
+
+This, in the cases where the type of `this` is not a template, maps to the
+existing practice of putting cv-ref qualifiers (that really apply to `*this` in
+current parlance) on the function itself, by writing them behind the parameter
+list, as far as the function signature is concerned.
+
+```cpp
+// Within a class
+
+struct Person {
+  std::string name;
+  
+  // Getter:
+  template <typename T>
+  decltype(auto) GetName(T&& this) { 
+      // T can only deduce to Person (mod cvref) because of visibility
+    return std::forward<T>(this).name;
+  }
+
+  // This template may instantiate the following functions:
+  // const std::string& GetName() const &;
+  // std::string& GetName() &;
+  // std::string&& GetName() &&;
+};
+
+void foo()
+{
+  string s = "Hello, world!";
+  
+  // Here, the closure object is an rvalue:
+  // the captured s will be correctly moved
+  bar([s] (auto&& this) { 
+    return std::forward<decltype(this)>(this).s;
+  }
+}
+```
+
+What does `this` in a parameter list mean?
+------------------------------------------
+
+The meaning of the different ways to pass `this` is very similar to the way
+`catch` works. This table uses `T` to refer to the enclosing class. `T` is not a
+`template`. This shall be illustrated on the method family `void f(...)...`.
```

Comment:
T is not a template _parameter_ I think is more accurate to what we want to say?
Maybe we should use Foo instead of T here, T is so ingrained as a template parameter, where Foo is more often used as a concrete class type.

Reply:
agree.

Reply:
Rephrased. done.